File: packages/react-native-sandbox/ios/SandboxRegistry.cpp
#include "SandboxRegistry.h"
#include <fmt/format.h>
#include <iostream>

namespace facebook {
namespace react {

/**
 * Implementation of the thread-safe C++ SandboxRegistry.
 *
 * This class provides a singleton registry for managing sandbox delegates
 * across multiple React Native instances. It uses std::recursive_mutex for
 * thread safety and std::map for efficient lookups.
 *
 * Key features:
 * - Thread-safe operations with recursive mutex
 * - Efficient O(log n) lookups using std::map
 * - Automatic cleanup of unregistered sandboxes
 * - Support for permission-based communication
 */

SandboxRegistry& SandboxRegistry::getInstance() {
  static SandboxRegistry instance;
  return instance;
}

void SandboxRegistry::registerSandbox(
    const std::string& origin,
    std::shared_ptr<ISandboxDelegate> delegate,
    const std::set<std::string>& allowedOrigins) {
  if (origin.empty() || !delegate) {
    return;
  }

  std::lock_guard<std::recursive_mutex> lock(registryMutex_);

  if (sandboxRegistry_.find(origin) != sandboxRegistry_.end()) {
    std::cerr
        << fmt::format(
               "[SandboxRegistry] Warning: Overwriting existing sandbox with origin: {}",
               origin)
        << std::endl;
  }

  sandboxRegistry_[origin] = delegate;
  allowedOrigins_[origin] = allowedOrigins;
}

void SandboxRegistry::unregister(const std::string& origin) {
  if (origin.empty()) {
    return;
  }

  std::lock_guard<std::recursive_mutex> lock(registryMutex_);

  auto registryIt = sandboxRegistry_.find(origin);
  if (registryIt != sandboxRegistry_.end()) {
    sandboxRegistry_.erase(registryIt);
  }

  auto originsIt = allowedOrigins_.find(origin);
  if (originsIt != allowedOrigins_.end()) {
    allowedOrigins_.erase(originsIt);
  }
}

std::shared_ptr<ISandboxDelegate> SandboxRegistry::find(
    const std::string& origin) {
  if (origin.empty()) {
    return nullptr;
  }

  std::lock_guard<std::recursive_mutex> lock(registryMutex_);

  auto it = sandboxRegistry_.find(origin);
  if (it != sandboxRegistry_.end()) {
    return it->second;
  }

  return nullptr;
}

bool SandboxRegistry::isPermittedFrom(
    const std::string& sourceOrigin,
    const std::string& targetOrigin) {
  if (sourceOrigin.empty() || targetOrigin.empty()) {
    return false;
  }

  std::lock_guard<std::recursive_mutex> lock(registryMutex_);

  auto originsIt = allowedOrigins_.find(sourceOrigin);
  if (originsIt == allowedOrigins_.end()) {
    return false;
  }

  return originsIt->second.find(targetOrigin) != originsIt->second.end();
}

void SandboxRegistry::reset() {
  std::lock_guard<std::recursive_mutex> lock(registryMutex_);
  sandboxRegistry_.clear();
  allowedOrigins_.clear();
}

} // namespace react
} // namespace facebook

File: packages/react-native-sandbox/android/src/main/java/io/callstack/rnsandbox/SandboxRegistry.kt
package io.callstack.rnsandbox

import android.util.Log

/**
 * Thread-safe singleton registry for managing sandbox delegates.
 * Mirrors the C++ SandboxRegistry for cross-sandbox communication.
 */
object SandboxRegistry {
    private const val TAG = "SandboxRegistry"
    private val lock = Any()
    private val sandboxRegistry = mutableMapOf<String, SandboxReactNativeDelegate>()
    private val allowedOriginsMap = mutableMapOf<String, Set<String>>()

    fun register(
        origin: String,
        delegate: SandboxReactNativeDelegate,
        allowedOrigins: Set<String>,
    ) {
        if (origin.isEmpty()) return

        synchronized(lock) {
            if (sandboxRegistry.containsKey(origin)) {
                Log.w(TAG, "Overwriting existing sandbox with origin: $origin, allowedOrigins=$allowedOrigins")
            } else {
                Log.d(TAG, "Registering sandbox origin: $origin, allowedOrigins=$allowedOrigins")
            }
            sandboxRegistry[origin] = delegate
            allowedOriginsMap[origin] = allowedOrigins
        }
    }

    fun unregister(origin: String) {
        if (origin.isEmpty()) return

        synchronized(lock) {
            sandboxRegistry.remove(origin)
            allowedOriginsMap.remove(origin)
        }
    }

    fun find(origin: String): SandboxReactNativeDelegate? {
        if (origin.isEmpty()) return null

        synchronized(lock) {
            return sandboxRegistry[origin]
        }
    }

    fun isPermittedFrom(
        sourceOrigin: String,
        targetOrigin: String,
    ): Boolean {
        if (sourceOrigin.isEmpty() || targetOrigin.isEmpty()) return false

        synchronized(lock) {
            val origins = allowedOriginsMap[sourceOrigin] ?: return false
            val permitted = origins.contains(targetOrigin)
            if (!permitted) {
                Log.w(TAG, "isPermittedFrom($sourceOrigin -> $targetOrigin): denied. allowedOrigins[$sourceOrigin]=$origins")
            }
            return permitted
        }
    }

    fun reset() {
        synchronized(lock) {
            sandboxRegistry.clear()
            allowedOriginsMap.clear()
        }
    }
}
